packages/flutter/lib/src/widgets/async.dart

```diff
@@ -372,42 +403,87 @@ typedef AsyncWidgetBuilder<T> = Widget Function(BuildContext context, AsyncSnaps
 class StreamBuilder<T> extends StreamBuilderBase<T, AsyncSnapshot<T>> {
   /// Creates a new [StreamBuilder] that builds itself based on the latest
-  /// snapshot of interaction with the specified [stream] and whose build
+  /// snapshot of interaction with the specified `stream` and whose build
   /// strategy is given by [builder].
   ///
-  /// The [initialData] is used to create the initial snapshot.
+  /// The [initialData] argument is used to create the initial snapshot.
   ///
   /// The [builder] must not be null.
   const StreamBuilder({
     Key key,
-    this.initialData,
+    @required T initialData,
+    Stream<T> stream,
+    @required this.builder,
+  }) : assert(builder != null),
+       hasInitialData = true,
+       _initialData = initialData,
+       super(key: key, stream: stream);
+
+  /// Creates a new [StreamBuilder] that builds itself based on the latest
+  /// snapshot of interaction with the specified `stream` and whose build
+  /// strategy is given by [builder].
+  ///
+  /// The initial snapshot will contain no data.
+  ///
+  /// The [builder] must not be null.
+  const StreamBuilder.withoutInitialData({
+    Key key,
     Stream<T> stream,
     @required this.builder,
   }) : assert(builder != null),
+       hasInitialData = false,
+       _initialData = null,
        super(key: key, stream: stream);
 
   /// The build strategy currently used by this builder.
   final AsyncWidgetBuilder<T> builder;
 
+  /// Whether this builder's initial snapshot contains data.
+  ///
+  /// If this is false, then attempting to access [initialData] will throw an
+  /// error.
+  ///
+  /// See also:
+  ///
+  ///  * [AsyncSnapshot.hasData], the corresponding property that will be set
+  ///    in the initial snapshot.
+  final bool hasInitialData;
+
   /// The data that will be used to create the initial snapshot.
   ///
   /// Providing this value (presumably obtained synchronously somehow when the
   /// [Stream] was created) ensures that the first frame will show useful data.
-  /// Otherwise, the first frame will be built with the value null, regardless
-  /// of whether a value is available on the stream: since streams are
-  /// asynchronous, no events from the stream can be obtained before the initial
-  /// build.
-  final T initialData;
+  /// Otherwise, the first frame will be built with a snapshot that contains no
+  /// data, regardless of whether a value is available on the stream: since
+  /// streams are asynchronous, no events from the stream can be obtained
+  /// before the initial build.
+  ///
+  /// Some builders intentionally have no data when first built. For those
+  /// cases, callers can use the [new StreamBuilder.withoutInitialData]
+  /// constructor. When a builder was constructed in this way, attempting to
+  /// access the [initialData] property will throw a [StateError].
+  T get initialData {
+    if (!hasInitialData)
+      throw StateError('No initial data');
```

Comment:
can we have a more useful error message?

Reply:
Would it be more appropriate to use `FlutterError` here, or is `StateError` better?  The docs on `FlutterError` say "Error class used to report Flutter-specific assertion failures and contract violations," but it's not clear to me from that description if widget implementations are encouraged to throw `FlutterError` or not.